subsys/fs/fs.c

```diff
@@ -0,0 +1,529 @@
+/*
+ * Copyright (c) 2018 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include <zephyr/types.h>
+#include <errno.h>
+#include <init.h>
+#include <fs.h>
+
+#define SYS_LOG_LEVEL CONFIG_SYS_LOG_FS_LEVEL
+#include <logging/sys_log.h>
+
+/* list of mounted file systems */
+static sys_dlist_t fs_mnt_list;
+
+/* lock to protect mount list operations */
+static struct k_mutex mutex;
+
+/* file system map table */
+static const fs_file_system_t *fs_map[FS_TYPE_END];
+
+int get_mnt_point(fs_mount_t **mnt_pntp, const char *name, int *match_len)
+{
+	fs_mount_t *mnt_p = NULL, *itr;
+	size_t longest_match = 0;
+	size_t len, name_len = strlen(name);
+	sys_dnode_t *node;
+
+	k_mutex_lock(&mutex, K_FOREVER);
+	SYS_DLIST_FOR_EACH_NODE(&fs_mnt_list, node) {
+		itr = CONTAINER_OF(node, fs_mount_t, node);
+		len = itr->mountp_len;
+
+		/*
+		 * Move to next node if mount point length is
+		 * shorter than longest_match match or if path
+		 * name is shorter than the mount point name.
+		 */
+		if ((len < longest_match) || (len > name_len)) {
+			continue;
+		}
+
+		/*
+		 * Move to next node if name does not have a directory
+		 * separator where mount point name ends.
+		 */
+		if ((len > 1) && (name[len] != '/') && (name[len] != '\0')) {
+			continue;
+		}
+
+		/* Check for mount point match */
+		if (strncmp(name, itr->mnt_point, len) == 0) {
+			mnt_p = itr;
+			longest_match = len;
+		}
+	}
+	k_mutex_unlock(&mutex);
+
+	if (mnt_p == NULL) {
+		return -ENOENT;
+	}
+
+	*mnt_pntp = mnt_p;
+	*match_len = mnt_p->mountp_len;
+	return 0;
+}
+
+/* File operations */
+int fs_open(fs_file_t *zfp, const char *file_name)
+{
+	fs_mount_t *mp;
+	const char *fs_path;
+	int rc, match_len;
+
+	if ((strlen(file_name) <= 1) || (file_name[0] != '/')) {
```

Comment:
The idea is based on the mount point (/fatfs or/nffs) through these mount points only we are finding the mounted file systems and calling the right fs interface function. So when we create a file or open a file we need to give the absolute path.

Reply:
That I understood. Why a mount point must have `/` at beginning?

Reply:
This will help us draw the similarity to Unix like files and I was thinking we can have pseudo root file system on '/' in future based on need.

Also it will be more clear if an absolute path starts with '/' than simply a name, easy for reading and debugging purposes.

Reply:
I would prefer for paths to begin with `/`; otherwise, it's going to be difficult to have a concept of "current working directory".